src/providers/postgres/qgspostgresprovider.cpp

```diff
@@ -3895,117 +3895,92 @@ QgsBox3D QgsPostgresProvider::extent3D() const
+  // Compute the extents, if estimation failed or was disabled
+  if ( ext.isEmpty() )
+    do
     {
       sql = QStringLiteral( "SELECT %1(%2%3) FROM %4%5" )
-            .arg( connectionRO()->majorVersion() < 2 ? "extent" : "ST_3DExtent",
+            .arg( vmaj < 2 ? "extent" : "ST_3DExtent",
                   quotedIdentifier( mBoundingBoxColumn ),
                   ( mSpatialColType == SctPcPatch || mSpatialColType == SctGeography ) ? "::geometry" : "",
                   mQuery,
                   filterWhereClause() );
 
       result = connectionRO()->LoggedPQexec( "QgsPostgresProvider", sql );
-      if ( result.PQresultStatus() != PGRES_TUPLES_OK )
-        connectionRO()->LoggedPQexecNR( "QgsPostgresProvider", QStringLiteral( "ROLLBACK" ) );
-      else if ( result.PQntuples() == 1 && !result.PQgetisnull( 0, 0 ) )
-        ext = result.PQgetvalue( 0, 0 );
+      if ( result.PQresultStatus() != PGRES_TUPLES_OK ) break;
+      if ( result.PQntuples() != 1 ) break;
+      if ( result.PQgetisnull( 0, 0 ) ) break;
+      ext = result.PQgetvalue( 0, 0 );
     }
+    while ( 0 );
 
-    if ( !ext.isEmpty() )
-    {
-      QgsDebugMsgLevel( QStringLiteral( "Got extents (%1) using: %2" ).arg( ext ).arg( sql ), 2 );
+  if ( !ext.isEmpty() )
+  {
+    QgsDebugMsgLevel( QStringLiteral( "Got extents (%1) using: %2" ).arg( ext ).arg( sql ), 2 );
 
-      const thread_local QRegularExpression rx3d( "\\((.+) (.+) (.+),(.+) (.+) (.+)\\)" );
-      const QRegularExpressionMatch match = rx3d.match( ext );
+    const thread_local QRegularExpression rx3d( "\\((.+) (.+) (.+),(.+) (.+) (.+)\\)" );
+    const QRegularExpressionMatch match = rx3d.match( ext );
+    if ( match.hasMatch() )
+    {
+      mLayerExtent.setXMinimum( match.captured( 1 ).toDouble() );
+      mLayerExtent.setYMinimum( match.captured( 2 ).toDouble() );
+      mLayerExtent.setZMinimum( match.captured( 3 ).toDouble() );
+      mLayerExtent.setXMaximum( match.captured( 4 ).toDouble() );
+      mLayerExtent.setYMaximum( match.captured( 5 ).toDouble() );
+      mLayerExtent.setZMaximum( match.captured( 6 ).toDouble() );
+    }
+    else
+    {
+      const thread_local QRegularExpression rx2d( "\\((.+) (.+),(.+) (.+)\\)" );
+      const QRegularExpressionMatch match = rx2d.match( ext );
       if ( match.hasMatch() )
       {
         mLayerExtent.setXMinimum( match.captured( 1 ).toDouble() );
         mLayerExtent.setYMinimum( match.captured( 2 ).toDouble() );
-        mLayerExtent.setZMinimum( match.captured( 3 ).toDouble() );
-        mLayerExtent.setXMaximum( match.captured( 4 ).toDouble() );
-        mLayerExtent.setYMaximum( match.captured( 5 ).toDouble() );
-        mLayerExtent.setZMaximum( match.captured( 6 ).toDouble() );
-
-        if ( !elevationProperties()->containsElevationData() )
-        {
-          mLayerExtent.setZMinimum( std::numeric_limits<double>::quiet_NaN() );
-          mLayerExtent.setZMaximum( std::numeric_limits<double>::quiet_NaN() );
-        }
```

Comment:
I should have let a comment on this part to explain the reason. As I remember that was necessary: sometimes the layer is viewed as a 2D layer even if it has 3D values in the extent request. This is why I set the Z coordinates to NaN.

Reply:
I just moved it outside the conditional, we're still setting it to NaN.
What it isn't clear is what the function should do when the extent could not be computed. Right now only Z min/max are set but all other dimensions are left untouched. Would it make more sense to setNull as the first thing ?

Reply:
With commit aa5e200d254a6acc4627d568d765f6bb1e333b84 I hopefully made it more clear too